### ISSUE TITLE
Solidjs hydration do not reevaluate conditional children following navigation with `ViewTransition`

### ISSUE THREAD

**Issue by mishimalisa** [withastro/astro#8781]
### Astro Info

```block
Astro                    v3.2.3
Node                     v18.18.0
System                   Linux (x64)
Package Manager          unknown
Output                   static
Adapter                  none
Integrations             @astrojs/solid-js
```

### If this issue only occurs in one browser, which browser is a problem?

Chrome

### Describe the Bug

When you use a global store or a signal with `<ViewTransitions />`, the hydration process doesn't reevaluate previous states that were used to determine what should be displayed in the children elements. However, the parent component continues to update its attributes even after you change the state and navigate away.

### What's the expected result?

I expected the store to maintain its state across navigations, which it does. However, I also anticipated that every rehydrated component would be re-evaluated according to the current store state. There are potential workarounds, but I suspect it's a hydration issue.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-bxb9uz-urtz5p?file=src%2Fpages%2Findex.astro,src%2Fcomponents%2FSolidItem.tsx&on=stackblitz

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by mishimalisa** [comment]
After upgrading to Solid `1.8`, I've realized that the HTML attributes no longer get updated, contrary to what I mentioned earlier. Now the problem is simpler: The component does not recover its modified state when it is unmounted and then remounted navigating using view transitions.

**Comment by martrapp** [comment]
Hi @mishimalisa,
I'll try to answer, even though I have no practical experience with solid-js.
View transitions are a valid approach to preserve the state of loaded JavaScript modules across navigation until the next full page reload.
The view you see when you return to the home page is the static view of the page with the initial state.
The components are reinserted into the DOM when you navigate to the view, but they don't recognize that they should redraw themselves according to the current state.   

The following changes both worked for me:
- Replace `client:load` with `client:only` or
- Add `onMount(() => setSignal((el) => ({ ...el })));` to the component with `onMount` from `'solid-js'`

**Comment by mishimalisa** [comment]
Hey @martrapp thanks for the comment,

> The view you see when you return to the home page is the static view of the page with the initial state.

The weird thing is, I noticed the "disabled" attribute doesn't revert to its initial static state when returning to the home page. [I updated the reproduction code to illustrate this](https://stackblitz.com/edit/github-bxb9uz-urtz5p?file=src%2Fpages%2Findex.astro,src%2Fcomponents%2FSolidItem.tsx&on=stackblitz).

I also tried this using Preact + nanostores. Interestingly, it does the exact opposite of what SolidJS does; it doesn't update the disabled attribute but does update its children based on the "modified" state after view transition navigation.

Preact + nanostores repro: https://stackblitz.com/edit/github-bxb9uz-gnd2zk?file=src%2Fcomponents%2FItem.tsx

**Comment by lilnasy** [comment]
I would not expect components to be "remounted" on a navigation. Although the cases brought up make me think there is inconsistency here.

**Comment by martrapp** [comment]
Yes @mishimalisa, I also noticed that `disabled` works somehow. And you had mentioned that in your first post. I have no idea how to explain this :). 

Technically, the state is not part of the component, but part of the module loader of the browser. The main difference when using View Transitions is that the module loader state is not cleared when navigating, unlike when fully loaded. This is good for maintaining state across pages in your example, but it seems to cause problems when it comes to hydration or other ESM-related things. You can't just run the same module again.  There doesn't seem to be a way to interact with the browser's module loader and unload/reload modules.  The JavaScript code of the astro-island custom component is executed on hydration, even when view transitions are used, but it does not seem to behave the same way as on reload.

**Comment by mishimalisa** [comment]
I see, it looks like the issue is that the componentes don't fully recognize the changes that happened after they were remounted for the second time due to the module not reloading.

Using Solid, a workaround I found is to create a signal inside the component, and update it using `createEffect`, which runs on mount and whenever the store changes.

Like this:
```jsx
function ShoppingCart() {
  const [state, setState] = createSignal();
  createEffect(() => setState(store.shoppingCartIsFull));
  return <div>Your shopping cart is {state() ? 'FULL' : 'EMPTY'}</div>;
}
```

I think anyone attempting to maintain state across navigations will come across this issue.
[I made another example to illustrate the issue and the workaround in a simple way](https://stackblitz.com/edit/github-bxb9uz-xgqnrk?file=src%2Fcomponents%2FShoppingCart.tsx).

**Comment by martrapp** [comment]
Hey @mishimalisa, thank you very much for your concise example including workaround!

**Comment by mishimalisa** [comment]
Yes, @martrapp. I think, as you said, the server is returning the static view ("rendered using the empty store"), and the hydrate function might be avoiding unnecessary work (such as DOM updates) because it trusts that the server's generated code represents the current correct state.

### Solution 1:
I think, using `transition:persist` is the preferred approach in most cases, but if you have a bunch of components on just one page (e.g.: "Add to shopping cart" button) that use a shared store, you might need those components on every page, plus the logic for showing and hiding the component visibility.

### Solution 2:
Using `client:only` works fine, but unfortunately the component is not visible before it loads.

### Solution 3:
[My solution works](https://github.com/withastro/astro/issues/8781#issuecomment-1783681098), but I think would be nice to have a more intuitive approach. One idea is to have a directive that replaces the static rendered component with a brand new client only component, like `<Button client:swap />`.

### Idea for a new hydration option for the UI frameworks
The combination of View Transitions API + Islands + Global Store is awesome, but might need new hydration options to recover shared states across navigations, so here is one idea: a new hydration directive that hydrates the component by treating the SSR HTML as stale and revalidates all its states.

```jsx
<Button hydrate:revalidate client:load />
```

So Instead of `hydrate(<Button />, node)`, we use `hydrate(<Button />, node, { mode: "revalidate" })`.

**Comment by lilnasy** [comment]
That seems like a good solution!

I say that because from what I've seen so far, solid intentionally tries _not_ to change server-rendered elements during hydration. The fact that it changes the `disabled` seems to be an oversight in their compiler.

I'm not sure about this yet, though. We will be confirming with the Solid team next week.

**Comment by mishimalisa** [comment]
In addition to that, If possible, Astro could provide options for hydrating (revalidating) after swap and before the view transition finishes. Which solves the problem of Flash of components not yet fully hydrated.

On first load it respects the `client:visible`. If a swap happens this component are hydrated (revalidating) after swap:
```jsx
<Button refresh:swap client:visible />
```

It always respects the `client:visible`:
```jsx
<Button refresh:client client:visible />
```

**Comment by lilnasy** [comment]
Directives would be a lot of effort because they will involve the compiler and documentation. Also, maybe there isn't enough reason to introduce a new concept, especially if the current use-case can be handled automatically.

For example, pretending all components are `client:only` would probably be enough. The idea being, to use the normal render function instead of the hydrate function, which solid-js seems to have unsafely optimized.

**Comment by mishimalisa** [comment]
I don't think solid-js is the only one that optimizes hydration.
I ran some tests to compare popular frameworks using their shared state techniques, to see if they can correctly hydrate the component that uses a shared store (mutated on the client only). I found  **Lit** and **Svelte** is the only frameworks that hydrates the mismatch correctly and doesn't emits an error.

| Framework | Updates children | Updates attributes | Emits hydration error |
|-----------|---------------------|-----------------------|-----------------|
| Preact    | ✅ Yes               | ❌ No                  | No        |
| Solid     | ❌ No                | ❌ No                  | No        |
| React     | ✅ Yes               | ✅ Yes                 | Yes (error)           |
| Lit       | ✅ Yes               | ✅ Yes                 | No        |
| Svelte    | ✅ Yes               | ✅ Yes                 | No        |
| Vue       | ✅ Yes               | ❌ No                  | Yes  (error)         |

Code: https://stackblitz.com/edit/github-bxb9uz-cwygvz?file=src%2Fpages%2Findex.astro

> Also, maybe there isn't enough reason to introduce a new concept, especially if the current use-case can be handled automatically.

I agree that `client:only` directive can suffice for most projects already using Astro. However, `client:only`  may lead to challenges reminiscent of "flash of white while using dark theme" issue and FOUC problem, which Astro has already resolved. I think it would be great to consider similar solutions for components (not present in every page) that share states as well.

>  The idea being, to use the normal render function instead of the hydrate function

Perhaps, Astro could fake a hydration by calling `render` somewhere and swapping it with the static component inside the `startViewTransition` function, so the enduser doesn't notice the flash?

**Comment by lilnasy** [comment]
> I don't think solid-js is the only one that optimizes hydration.

You're right. This issue is relevant to all frameworks.

While the `disabled` attribute updating on navigation is still a minor bug for Solid.js, `ryansolid` confirmed that hydration is not supposed to perform a DOM mutation, and also that hydration mismatches should not exist in the first place. [Solid.js discord](https://discord.com/channels/722131463138705510/722131463889223772/1168790836687929385).

This is a problem for astro islands to solve, and I think it is approachable one for the reason you mentioned: we are free to perform any DOM operation before we show something to the visitor, no flashes necessary.

All this means, this a new feature request: to figure out a story for state-persisting client components with view transitions.

**Comment by martrapp** [comment]
We are already doing some pre-rendering to determine the stylesheets of cliet-only components during view transitions. I like the idea of updating the static view for non-persisted islands, as this should also eliminate the hydration errors we should expect from Vue, for example.

**Comment by mishimalisa** [comment]
In detail: Astro can use `render` function instead of `hydrate` for each desired island on the next navigated page, and replace the generated static island from the server with the client-rendered island during the pre-rendering phase (inside `startViewTransition`). The main issue is that every component on the next page must be downloaded and rerendered before loading the next page (which I think is fine, is basically how metaframeworks usually work). The worst case that I can think of is a scenario where most components that want to preserve their state actually did not change their state, which makes the use of `render` unnecessary extra work. This should be opt-in per component. Also note that the `render` would happen only after navigation not on first load.

In short when navigation happens:
1. Fetch the next page
2. Find any element that may require `render`
3. Fetch the components that require `render`
4. Meanwhile, pre-render the static page
5. When the fetched components are ready replace static islands with the rendered islands
6. Finish the view transition

A more adventurous approach would be caching a reference for each unique island that the end user has seen/loaded, just like `transition:persist`, but persisting forever somewhere. The issue is that the user might update the shared store before components relying on it are cached, leading to a mismatch later on. The caching might be helpful for simple use cases where the component updates only affect itself.

Regarding hydration,
[From svelte docs about hydration:](https://svelte.dev/docs/client-side-component-api#creating-a-component)
> The existing DOM doesn't need to match the component — Svelte will 'repair' the DOM as it goes.


I also [tried to break Svelte and Lit hydration](https://stackblitz.com/edit/github-bxb9uz-scht26?file=src%2Fcomponents%2FSvelte.svelte), without success. Interestingly seems like Svelte knows how to reconcile any mismatch during hydration; the same goes for Lit.

**Comment by lilnasy** [comment]
You mention that classic render would be unnecessary extra work. I think that's technically true but I would expect the extra work to be on the order of a millisecond. Do you have special requirements where you know that's not the case?

**Comment by mishimalisa** [comment]
@lilnasy I agree it should be fast enough for all use cases and I can't think of a best solution. The cost is negligible, but it is still something to be aware of. If some components decide to block for whatever reason it can make the experience worse than a normal navigation.

**Comment by mishimalisa** [comment]
I started a proposal about this here: https://github.com/withastro/roadmap/discussions/756

**Comment by peerreynders** [comment]
> While the `disabled` attribute updating on navigation is still a minor bug for Solid.js

I suspect there is something more significant going on here. 

I essentially arrived at what I would consider a reasonable solution:
```TypeScript
// file: src/components/SolidItem.tsx
import { createMemo, createSignal, onMount } from 'solid-js';
import { accessor, setter, type Core } from '../store';
import type { Accessor } from 'solid-js';

const toggleById = (cache: Core, id: string) => ({
  ...cache,
  [id]: !Boolean(cache[id]),
});

const nextSync = (state: number) => state + 1;
const derived = (
  accessor: Accessor<Core>,
  id: string,
  sync: Accessor<number>
) =>
  accessor()[id]
    ? {
        class: 'bt bt--disabled active',
        disabled: true,
        content: 'Active ❤️',
        signature: 'active disabled',
        sync: sync(),
      }
    : {
        class: 'bt idle',
        disabled: false,
        content: 'Like button!',
        signature: 'idle enabled',
        sync: sync(),
      };
/* https://css-tricks.com/making-disabled-buttons-more-inclusive */

export default function SolidItem(props: { id: string }) {
  let button: HTMLButtonElement | undefined;
  // sync() only exists to trigger re-evaluation of the memo
  // when server rendered component needs to be synchronized
  // with divergent shared client state.
  const [sync, setSync] = createSignal(Number.MIN_SAFE_INTEGER);
  const state = createMemo(() => derived(accessor, props.id, sync));
 
  onMount(function syncWithClientSharedState() {
    // Compare the server rendered signature with the
    // transitioned client state signature
    // if (state().signature === button.dataset.signature) return;
    if (state().class === button.className) return;

    // Synchronize if divergent
    // console.log('sync', props.id, button.dataset.signature, state().signature);
    console.log('sync', props.id, button.className, state().class);
    setSync(nextSync);
  });

  const toggle = () => {
    if (state().disabled) return;
    setter((cache) => toggleById(cache, props.id));
  };

  return (
    <button
      ref={button}
      type="button"
      onClick={toggle}
      class={state().class}
    >
      <span class="front">{state().content}</span>
    </button>
  );
}
```
https://stackblitz.com/edit/github-bxb9uz-rozfzs?file=src%2Fcomponents%2FSolidItem.tsx

However this breaks once I add more attributes to the `<button>` element:
```TypeScript
  return (
    <button
      ref={button}
      type="button"
      onClick={toggle}
      class={state().class}
      aria-disabled={state().disabled}
      data-signature={state().signature}
    >
      <span class="front">{state().content}</span>
    </button>
   );
```

It seems that those attributes have lost their reactivity with regards to programmatic updates (the `content` updates just fine). 

When I remove the `disabled` guard, user interaction seems to restore reactivity but before that point any amount of programmatic prodding of the `memo` doesn't seem to propagate the correct attribute values in the DOM.

**Comment by danielo515** [comment]
So shared persistent global state is still a problem to solve except for svelte? If I only use astro and svelte components does this work?

**Comment by victor-auffret** [comment]
With solidjs, you need juste extract createSignal
`
import { createSignal } from "solid-js";

const [count, setCount] = createSignal(0);

const Counter= () => {
  const increment = () => setCount((c) => c + 1);
  return <button onclick={increment}> compteur : {count()} </button>;
};

export { Counter};
`
So, the state is share between all component. If you need a "persist state" you just need state out of component but if you need state per component maybe you need something else

**Comment by Arecsu** [comment]
I believe I managed to solve this in a mostly Astro way 😄 
I didn't touch Astro's Framework code. Just played with `<ClientRouter />` events to get this working.

**Here is the repo with the implementation** 
https://github.com/Arecsu/astro-transition-swap-demo

**Here is the demo**
https://arecsu.github.io/astro-transition-swap-demo/

This is a related proposal to the issue: https://github.com/withastro/roadmap/discussions/756

It discusses the creation of a `client:swap` directive, but I think this should belong to be part of the `transition` set of directives, given it needs the SPA-like routing `<ClientRouter />` provides to preserve states between route changes. Could become `transition:swap` if officially implemented, I guess.

What this does in a nutshell:

1. We tag the components with a given "data-swap-id". In this case, the line in `index.astro` becomes:

```astro
<Counter client:load data-swap-id="demo-counter" />
``` 

Again, I'm using "data-swap-id" just for the purposes of this hacky demo.

2. `astroClientSwap.ts` has the crucial code for this. On route changes, it will look for elements with these "data-swap-id" and it will move them to a hidden element in the `<body>` that has `transition:persist`, hopefully in the core layout being used across all pages to be preserved. 

The code in Layout.astro becomes: 

```astro
    <!-- This is where the astro-island components will be saved across route changes -->
    <div
      id="client-swap-container"
      transition:persist
      style="display: none !important; position: absolute !important; top: -9999px !important; left: -9999px !important; visibility: hidden !important; pointer-events: none !important;"
    ></div>
```

In an official Astro implementation this might be taken care of automatically with no need to create this element.

3. After page swaps, it will look for `<astro-island>` components that have the same `data-swap-id` as the ones that have been saved in this hidden element.
4. For each found component, swap them back in favor of the user's island. I'm targeting `<astro-island>` because that was my initial use-case with Svelte components. I believe `transition:` directives work with any HTML element though, so this might be something to be considered in an official implementation.

Performance wise, I don't think we can get better than this. It should preserve ALL javascript and DOM states. Not even external stores or CSS hide-until-mount are necessary to keep the data (which, still, didn't fix the issue because we had to deal with the initial data statically display by the un-hydrated astro-island). It feels very Astro UX and DX wise.

Previously, to do that, we would have to workaround a `transition:persist` element in common across all pages, which would become incredibly hard to keep up. This should handle it.

Thoughts on this?

**Comment by matthewp** [comment]
We're closing stale issues to keep the backlog manageable as the project has grown. Since this issue hasn't seen activity in a while, we're closing it out. If you're still experiencing this in Astro 6, please open a new issue and we'll take a fresh look. Thank you for reporting it!
